### PR TITLE
Memphis week2 ass3

### PR DESCRIPTION
**Pull request by memphis88**

### CODE REVIEW

File: memphis88/week2/ass3/index.html [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+	var AssignmentApp = React.createClass({
+		mixins: [
+			OnChangeMixin,
+			OnPressingEnterMixin
+		],
+
+		getInitialState: function() {
+			return {
+				generatedLists: [],
+				text: '',
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: maybe call this `listName`

File: memphis88/week2/ass3/index.html [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+		},
+
+		submitMessage: function(listKey, msgText, msgKey) {
+			if (msgText === '') { return; };
+			if (this.isValidMessage(msgText)) {	
+				var updatedList = this.state.generatedLists.map(function(list) {
+					if (list.index == listKey) {
+						list.messageList.push({
+							text: msgText,
+							index: msgKey.toString().includes('.') ? msgKey : listKey.toString() + '.' + msgKey.toString()
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: If you something unique, you wont have to maintain this.

File: memphis88/week2/ass3/index.html [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+				generatedLists: listsCopy,
+				text: '',
+				index: newIndex
+			});
+		},
+
+		submitMessage: function(listKey, msgText, msgKey) {
+			if (msgText === '') { return; };
+			if (this.isValidMessage(msgText)) {	
+				var updatedList = this.state.generatedLists.map(function(list) {
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: you can use a `filter` here to reduce nesting

File: memphis88/week2/ass3/index.html [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+				this.setState({
+					generatedLists: updatedList
+				});
+			} else {
+				//TODO handle error
+			}
+		},
+
+		deleteMessage: function(listKey, msgKey) {
+			var updatedList = this.state.generatedLists.map(function(list) {
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: use filters

File: memphis88/week2/ass3/index.html [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+				if (list.index == listKey) {
+					var updatedMessages = list.messageList.filter(function(msg) {
+						if (msg.index == msgKey) { return false; }
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: `return msg.index != msgKey;`

File: memphis88/week2/ass3/index.html [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+					list.archivedList = updatedArchive;
+					list.messageList.push(msgToExtract);
+				}
+				return list;
+			});
+			this.setState({ generatedLists: updatedList });
+		},
+
+		moveMessage: function(msg, oldListKey, targetListKey) {
+			if (oldListKey === 'undefined' || targetListKey === 'undefined' || msg === 'undefined') { return; }
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: this doesn't check for undefined

File: memphis88/week2/ass3/index.html [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+		}
+	});
+	var GeneratedList = React.createClass({
+		mixins: [
+			OnChangeMixin,
+			OnPressingEnterMixin
+		],
+
+		propTypes: {
+
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: :(

**Comment by memphis88** [comment]
It turns out that my model is very inconvenient to handle without mutating state. I'll look a little more if there are ways of doing it correctly without big nesting otherwise I'll consider changing the model.

**Comment by hulvei3** [comment]
thats fine. forget the state mutation for now. Reduce nesting and impl the other minor stuff we talked about

**Comment by memphis88** [comment]
I ended up using `JSON.stringify()` and `JSON.parse()` to make deep copies. I don't like this solution that much but it gives the most readable result without mutating state.

**Comment by memphis88** [comment]
Let me know if this is ready for merge so I can continue with the next assignment :-)

File: memphis88/week2/ass3/index.html [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+			var msgToExtract;
+			generatedList.archivedList = generatedList.archivedList.filter(function(message) {
+				if (message.index != msgKey) {
+					msgToExtract = message;
+					return false;
+				}
+				return true;
+			});
+			generatedList.messageList.push(msgToExtract);
+			this.setState({ generatedLists: updatedList });
```
Comment (hulvei3) [inline memphis88/week2/ass3/index.html]: this will not work (`updatedList`)

**Comment by hulvei3** [comment]
Very nice and clean(er) code! Great work!
Fix the undefined error and merge it yourself when done

**Pull request merged**